DS-3542 Spring security authorizations 2

tomdesair opened the pull request:
This PR replaces https://github.com/DSpace/DSpace/pull/1916

This PR integrates Spring Security authorizations checks with the new DSpace 7 REST API. We're still in the process of testing and improving this.

Commit:
DS-3542: Spring permission evaluator for REST API

tomdesair commented:
@abollini I realised that after the rebase we did not add authorization checks on the new REST repositories of the submission. Therefore I've added some TODO's in this commit https://github.com/DSpace/DSpace/pull/1976/commits/368c68eb1b1ca5b545a37873b956835b2b1eff6b

Since you are more familiar with the submission endpoints, can you check if those permission checks make sense? Do we implement them now or in another ticket?

terrywbrady commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/ItemRestRepository.java:
```diff
@@ -46,21 +48,24 @@ public ItemRestRepository() {
 
     @Override
-    public Page<ItemRest> findAll(Context context, Pageable pageable) {
+    @PreAuthorize("hasAuthority('ADMIN')")
```
Is findAll() a method that should only be accessed by an Admin?  Does findAll() ever mean "find all that I can access" or is that type of operation handled elsewhere?

tomdesair replied:
Currently the `findAll` method returns all items ignoring any resource policies. That is why only admins should be allowed to use this endpoint. (see the discussion in https://github.com/DSpace/DSpace/pull/1879)

terrywbrady commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/AuthorityEntryLinkRepository.java:
```diff
@@ -54,6 +54,7 @@ public HALResource wrapResource(AuthorityEntryRest model, String... rels) {
         return new AuthorityEntryResource(model);
     }
 
+    //TODO @PreAuthorize("hasAuthority('EPERSON')")
```
I see a couple annotations marked with TODO.  How will these TODOs be resolved?

terrywbrady commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/MetadataSchemaRestRepository.java:
```diff
@@ -53,10 +52,10 @@ public MetadataSchemaRest findOne(Context context, Integer id) {
 
     @Override
-    public Page<MetadataSchemaRest> findAll(Context context, Pageable pageable) {
+    public Page<MetadataSchemaRest> findAll(Pageable pageable) {
```
Question: for documentation purposes, should all endpoint methods have an authorization annotation (including ones allowed by anonymous) to make the permissions very explicit?

tomdesair replied:
Good point, however I think that people we easily forget the `@PreAuthorize("hasAuthority('ANONYMOUS')")` as it will have no effect.

terrywbrady commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/SiteRestRepository.java:
```diff
@@ -44,10 +43,10 @@ public SiteRestRepository() {
 
     @Override
-    public SiteRest findOne(Context context, UUID id) {
+    public SiteRest findOne(UUID id) {
```
Since one and only one site object will exist, should there be another method to pull the single site object without specifying the id?

tomdesair replied:
Is the `Site` object actually used in DSpace 6?

terrywbrady commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/security/AdminRestPermissionEvaluatorPlugin.java:
```diff
@@ -0,0 +1,69 @@
+    @Autowired
+    private EPersonService ePersonService;
+
+    public boolean hasPermission(Authentication authentication, Serializable targetId, String targetType,
```
Add javadoc to clarify purpose and format of targetType and permission.  Since this is an admin check, note that permission is unused.  Could an enum be used for permission?

terrywbrady commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/security/DSpacePermissionEvaluator.java:
```diff
@@ -0,0 +1,44 @@
+public class DSpacePermissionEvaluator implements PermissionEvaluator {
+
+    @Autowired
+    private List<RestPermissionEvaluatorPlugin> permissionEvaluatorPluginList;
```
What are some examples of different plugins that might be in use at one time?  Is this to support the notion of "special groups" from earlier versions of DSpace, or is this a new concept?

tomdesair replied:
This is a new concept. For example, in DSpace 6 there are no resource policies attached to an EPerson object. This means the normal `AuthorizeService` cannot decide if authenticated EPerson A is allowed to view or update data of EPerson B. 

During one of the developer meetings, we decided to solve this problem only in the REST layer for now. Later (DSpace 8 or 9), this "authorize/permission plugin" system has to be pushed down to the "dspace-api" level.

terrywbrady commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -54,32 +66,40 @@ public void findAllPaginationTest() throws Exception {
+                        .andExpect(jsonPath("$._embedded.epersons", Matchers.contains(Matchers.anyOf(
+                       EPersonMatcher.matchEPersonOnEmail(eperson.getEmail()),
+                       EPersonMatcher.matchEPersonOnEmail(admin.getEmail()),
+                       EPersonMatcher.matchEPersonOnEmail(ePerson.getEmail())
```
The variables eperson and ePerson are confusing in the same block.  Could more specific names be used?

terrywbrady commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -306,9 +309,220 @@ public void findOneTestWrongUUID() throws Exception {
 
+    @Test
+    public void embargoAnonymousAccessTest() throws Exception{
+        context.turnOffAuthorisationSystem();
```
should calls to turnOffAuthorizationSystem() be wrapped in a try... finally to ensure that context is always reset properly?  Or, is the context recreated with each test?

terrywbrady reviewed (commented):
These changes are thoughtfully implemented.  I have embedded a few questions that would be useful to discuss.

YanaDePauw committed:
DS-3542: Spring Authorization - @Preauthorization

terrywbrady commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/security/GroupRestPermissionEvaluatorPlugin.java:
```diff
@@ -0,0 +1,79 @@
+     * @return true if the permission is granted by one of the plugins, false otherwise
+     */
+    public boolean hasPermission(Authentication authentication, Serializable targetId,
+                                 String targetType, Object permission) {
```
The permission object is not used/evaluated.

terrywbrady commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/security/AdminRestPermissionEvaluatorPlugin.java:
```diff
@@ -0,0 +1,74 @@
+        Context context = ContextUtil.obtainContext(request.getServletRequest());
+        EPerson ePerson = null;
+        try {
+            ePerson = ePersonService.findByEmail(context, (String) authentication.getPrincipal());
```
Consider checking if ePerson is a repository admin.  If so, return true immediately.

terrywbrady reviewed (commented):
This code looks good to me.  Those code is clean and well-documented.

See the 2 comments that I have suggested.  I have not yet reviewed the test classes in detail.

I have not yet tested these changes.  Here are my next steps for the review.
- attempt to resolve the merge issue (done)
- test locally (verified the restriction on the /items service)
- perform a closer review of the test classes

#### Question: 
The restriction on the /items endpoint does ensure that when the method returns data it is returning data consistently. 

If someone is learning the API, it would be natural for them to want to run the /items endpoint as they are leaning the API.  We allow communties and collections to be retrieved.  As a new user of the API, I would expect similar behavior for items.

If we must restrict this API, could the error message direct the user to try /discover/browses/title/items?

terrywbrady commented:
@tomdesair , I hope you do not mind that I resolved the merge issue on your PR.

terrywbrady commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:
```diff
@@ -1660,7 +1660,7 @@ public void discoverSearchObjectsTestForEmbargoedItemsAndPrivateItems() throws E
                                       .withAuthor("Smith, Maria").withAuthor("Doe, Jane").withAuthor("Testing, Works")
                                       .withSubject("TestingForMore").withSubject("ExtraEntry")
-                                      .makePrivate()
+                                      .makeUnDiscoverable()
```
Consider renaming the method from "private" to "undiscoverable" to match the method names in use.

terrywbrady commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -306,9 +330,215 @@ public void findOneTestWrongUUID() throws Exception {
+        context.restoreAuthSystemState();
+
+
+        getClient().perform(get("/api/core/items/" + embargoedItem1.getID()))
```
Since publicItem1 was created, should a retrieve test be performed here to illustrate the successful return?

terrywbrady commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -306,9 +330,215 @@ public void findOneTestWrongUUID() throws Exception {
+        String token1 = getAuthToken(admin.getEmail(), password);
+
+
+        getClient(token1).perform(get("/api/core/items/" + privateItem1.getID()))
```
It would be good to illustrate the failure for an anonymous user attempting to access this item.

tomdesair replied:
Anonymous users are allowed to access "private/undiscoverable" items. I also think this is confusing but that's how it works in DSpace 6 at the moment.

terrywbrady reviewed (approved):
Following up to my prior review, I took a look at the test classes and had some very minor suggestions.

Note, in the tests for the /api/config/submissiondefinitions/ endpoint, valid submission and upload is performed by an admin user.  Would it be a better test to construct a user with submission privileges rather than an admin user?

tomdesair commented:
@abollini and @tdonohue This PR is ready to merge.

The PR adds authorizations checks at the REST API level using [Spring Authorization annotations](https://docs.spring.io/spring-security/site/docs/current/reference/html/el-access.html) (as discussed in the [Dev meeting of 2018-01-11](https://wiki.duraspace.org/display/DSPACE/2018-01-11+DSpace+7+Working+Group+Meeting+notes))

Currently we only make use of the `@PreAuthorize` annotations. This means Spring will evaluate the expression within the annotations before executing the annotated method. If the evaluation of the expression fails or returns false, Spring will not execute the method and return a 403.

The type of expressions we use are:

- `hasPermission(#uuid, 'DSO-TYPE', 'ACTION')`: Check if the current user is allowed to execute the listed action on the specified DSpace Object (for example [downloading a bitstream](https://github.com/DSpace/DSpace/pull/1976/files#diff-ac78d7e83f5c57bba55a5f0ab9d01418R73)).

- `hasAuthority('VALUE')`: Check if the current user has a specific Spring authority. Currently there are only three authority values: `ADMIN`, `EPERSON` and `ANONYMOUS`. (for example [when querying all items](https://github.com/DSpace/DSpace/pull/1976/files#diff-3a2e853733bc80822b73da931c8534c9R80)).

For the evaluation of the `hasPermission` expressions, we wrote a custom "permission evaluator" `DSpacePermissionEvaluator` that uses a plug-in system. If one of the available plug-ins approves the requested persmission, the current user is allowed to execute the action. The plug-ins we implemented are:

1. `AuthorizeServicePermissionEvaluatorPlugin`: Check persmissions based on the DSpace `AuthorizeService`. This service will validate if the authenticated user is allowed to perform an action on the given DSpace Object based on the resource policies attached to that DSpace object.

1. `EPersonRestPermissionEvaluatorPlugin`: An authenicated user is allowed to view, update or delete his or her own data. Since there are no explicit resource policies for this, the `AuthorizeService` does not cover this use case.

1. `GroupRestPermissionEvaluatorPlugin`: An authenticated user is allowed to view information on all the groups he or she is a member of (READ permission). Since there are no explicit resource policies for this, the `AuthorizeService` does not cover this use case.

1. `AdminRestPermissionEvaluatorPlugin`: Administrators are always allowed to perform any action on any DSpace object. This plugin will check if the authenticated EPerson is an administrator of the provided target DSpace Object. If that is the case, the authenticated EPerson is allowed to perform the requested action.

We've annotated the methods of which the authorizations requirements were clear for us. We've added integration tests to explicitly test again authorized and unauthorized access.

tdonohue commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/test/AbstractControllerIntegrationTest.java:
```diff
@@ -54,15 +55,16 @@
 
     protected static final String AUTHORIZATION_HEADER = "Authorization";
-    protected static final String AUTHORIZATION_TYPE = "Bearer";
+    protected static final String AUTHORIZATION_TYPE = "Bearer ";
```
Is the extra space here on purpose?  I'm guessing this is an accident.  If it is necessary, we should add a comment as to why it's necessary.

tdonohue commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/AuthorityEntryLinkRepository.java:
```diff
@@ -54,6 +55,7 @@ public HALResource wrapResource(AuthorityEntryRest model, String... rels) {
         return new AuthorityEntryResource(model);
     }
 
+    @PreAuthorize("hasAuthority('EPERSON')")
```
`hasAuthority('EPERSON')` is awkwardly named, in my opinion.  EPerson is a type of object in DSpace and not a permission level.  It's confusing to see, for example, `hasPermission(#id, 'EPERSON', 'READ')` and `hasAuthority('EPERSON')` alongside one another, as the word EPERSON has two different meanings (the first is a specific type of object, the second seems to be implying you are "authenticated" and don't necessarily have specific rights on an EPERSON object).

Should we consider renaming this to `hasAuthority('AUTHENTICATED')`?  If I understand this, that's the true meaning -- you simply need to be authenticated to access this endpoint.

tdonohue reviewed (approved):
👍  Overall, this code looks great! Has tests and javadocs, and seems like a good implementation.

That said, I noticed a few small things that I commented on below.  While I'm OK with this PR "as-is", I think these two things might be improved on.

Commit:
DS-3542: Renamed EPERSON Spring security grant to AUTHENTICATED

abollini commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -129,37 +138,47 @@ public void findAllWithPaginationTest() throws Exception {
+        String token = getAuthToken(admin.getEmail(), password);
+
+        getClient(token).perform(get("/api/core/items")
+                                         .param("size", "2"))
```
why we have a so large diff due to formatting style? @tdonohue we should improve our code style if different IDE format the code in so different way (making huge the diff) without violate our code style... can you take a look to that or open an issue on JIRA? @tomdesair are you willing to make some cleanup to reduce the changes?

tomdesair replied:
During the last rebase I just restored the original formatting which was changed by this PR https://github.com/DSpace/DSpace/pull/2064/files#diff-8bb61e42ed580192f717e4c1f8f94f4dR148

abollini reviewed (changes requested):
thanks @tomdesair for this large contribution. The code looks good I have only inspected it visually. I have found some misuse of the 403 Forbidden status vs 401 Unauthorized status code. Myself found anytime that statuses really confusing but the 403 should be returned only if the provided credentials are not sufficient to perform the action. If the user is not yet logged in a 401 Unauthorized must be returned (there is lot of discussion in the web with many saying that it should be named Unauthenticated... but that's it)

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamContentRestControllerIT.java:
```diff
@@ -306,7 +306,7 @@ public void testPrivateBitstream() throws Exception {
 
                        //** THEN **
-                       .andExpect(status().isUnauthorized());
+                       .andExpect(status().isForbidden());
```
This response code is set by the Spring Security framework and I don't find a way to configure another response code. So I'm afraid I can't change this with the current approach.

tomdesair commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java:
```diff
@@ -69,13 +69,6 @@
-    @Override
-    public T findOne(ID id) {
-        Context context = obtainContext();
-        return findOne(context, id);
```
The Spring Security annotations are based on Spring AOP. Spring AOP will enhance each annotated class with a facade. This facade will check each incoming method call from another object for the correct authorizations. However, method calls within the same class are not checked.

Since `org.dspace.app.rest.RestResourceController`always uses `findOne(ID id)` without `Context`, the only method that can be annotated is `findOne(ID id)`. Therefore we removed it from the super-class so that each sub-class can define his own custom permissions.

So an alternative approach to change `RestResourceController` to use `findOne(Context context, ID id)`. 

However, it is important that we only keep either `findOne(ID id)` **or** `findOne(Context context, ID id)` for the annotations to work.

tomdesair commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/utils/RestRepositoryUtils.java:
```diff
@@ -60,7 +61,9 @@
     public boolean haveSearchMethods(DSpaceRestRepository repository) {
         for (Method method : repository.getClass().getMethods()) {
-            SearchRestMethod ann = method.getAnnotation(SearchRestMethod.class);
+//            SearchRestMethod ann = method.getAnnotation(SearchRestMethod.class);
```
That approach does not work with the facade injected by Spring AOP (see above). I'll add a comment to that line.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/BitstreamRestRepositoryIT.java:
```diff
@@ -172,9 +174,10 @@ public void findAllPaginationTest() throws Exception {
 
-        ;
+        getClient().perform(get("/api/core/bitstreams/"))
+                .andExpect(status().isForbidden());
```
This response code is set by the Spring Security framework and I don't find a way to configure another response code. So I'm afraid I can't change this with the current approach.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -33,129 +37,151 @@ public void findAllTest() throws Exception {
         ;
+
+        getClient().perform(get("/api/eperson/epersons"))
+                   .andExpect(status().isForbidden())
```
This response code is set by the Spring Security framework and I don't find a way to configure another response code. So I'm afraid I can't change this with the current approach.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -33,129 +37,151 @@ public void findAllTest() throws Exception {
-                                        .withNameInMetadata("John", "Doe")
-                                        .withEmail("Johndoe@gmail.com")
-                                        .build();
+        ePersonService.setPassword(eperson, password);
```
This indeed looks a bit weird. I'll ask a developer to check.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:
```diff
@@ -33,129 +37,151 @@ public void findAllTest() throws Exception {
-                                         .withNameInMetadata("Jane", "Smith")
-                                         .withEmail("janesmith@gmail.com")
-                                         .build();
+        //Only works with admin
```
An authenticated user is able to only read his own data. See test below this one (`readEpersonAuthorizationTest`).

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java:
```diff
@@ -31,25 +31,41 @@
+
         getClient().perform(get("/api/eperson/groups"))
+                   //The status has to be 403 Not Authorized
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/GroupRestRepositoryIT.java:
```diff
@@ -31,25 +31,41 @@
-        context.turnOffAuthorisationSystem();
+
         getClient().perform(get("/api/eperson/groups"))
+                   //The status has to be 403 Not Authorized
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/SubmissionDefinitionsControllerIT.java:
```diff
@@ -34,8 +34,16 @@
+        //When we call the root endpoint as anonymous user
         getClient().perform(get("/api/config/submissiondefinitions"))
+                   //The status has to be 403 Not Authorized
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/SubmissionDefinitionsControllerIT.java:
```diff
@@ -80,6 +95,16 @@ public void findByCollection() throws Exception {
+                                    .param("uuid", col1.getID().toString()))
+                   //** THEN **
+                   //The status has to be 200
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security. We'll adjust the comment

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/SubmissionDefinitionsControllerIT.java:
```diff
@@ -93,8 +118,17 @@ public void findByCollection() throws Exception {
         //Match only that a section exists with a submission configuration behind
         getClient().perform(get("/api/config/submissiondefinitions/traditional/collections"))
+                   //The status has to be 403 Not Authorized
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/SubmissionDefinitionsControllerIT.java:
```diff
@@ -104,7 +138,14 @@ public void findCollections() throws Exception {
+
         getClient().perform(get("/api/config/submissiondefinitions/traditional/sections"))
+                   //The status has to be 403 Not Authorized
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/SubmissionFormsControllerIT.java:
```diff
@@ -28,8 +28,16 @@
+        //When we call the root endpoint as anonymous user
         getClient().perform(get("/api/config/submissionforms"))
+                   //The status has to be 403 Not Authorized
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/SubmissionFormsControllerIT.java:
```diff
@@ -51,7 +59,15 @@ public void findAll() throws Exception {
+        //When we call the root endpoint as anonymous user
         getClient().perform(get("/api/config/submissionforms/traditionalpageone"))
+                   //The status has to be 403 Not Authorized
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/SubmissionSectionsControllerIT.java:
```diff
@@ -27,8 +27,16 @@
+        //When we call the root endpoint as anonymous user
         getClient().perform(get("/api/config/submissionsections"))
+                   //The status has to be 403 Not Authorized
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/SubmissionUploadsControllerIT.java:
```diff
@@ -27,8 +27,16 @@
+        //When we call the root endpoint as anonymous user
         getClient().perform(get("/api/config/submissionuploads"))
+                   //The status has to be 403 Not Authorized
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -1033,4 +1063,193 @@ public void valueMissingForDiscoverableOperation() throws Exception {
+
+        //An anonymous user is not allowed to view embargoed items
+        getClient().perform(get("/api/core/items/" + embargoedItem1.getID()))
+                   .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented on dspace-spring-rest/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -1033,4 +1063,193 @@ public void valueMissingForDiscoverableOperation() throws Exception {
+
+        //An anonymous user is not allowed to the restricted item
+        getClient().perform(get("/api/core/items/" + restrictedItem1.getID()))
+                .andExpect(status().isForbidden());
```
Same as above, I did not find a way to configure this in Spring Security.

tomdesair commented:
Because the flood of comments doesn't really give a good overview, I'll summarize our TODO's here:

1. We did not find a way to change the response code returned by Spring Security from Forbidden (403) to Unauthorized (401). But please read also this discussion: https://stackoverflow.com/questions/3297048/403-forbidden-vs-401-unauthorized-http-responses

2. We'll cleanup and add comments to the usage of ClassUtils

3. If desired I can change `findOne(ID id)` to `findOne(Context context, ID id)` but we can keep only one of these methods with this approach.

Commit:
DS-3542: Added custom DSpace AuthenticationEntryPoint in order to return 401 status

tomdesair commented:
Hi @abollini,

I've implemented your first suggestion to use a custom `AuthenticationEntryPoint` and this restores the 401 status code for requests where the user did not provide an authentication token (see commit https://github.com/DSpace/DSpace/pull/1976/commits/2bc9b22e62592fb78f0290f517a75efe40021940). Is this OK?

I'll rebase and look at your other suggestion in the coming days.

Commit:
DS-3542: Correct DSO Admin level check

Commit:
DS-3542: Restore DSpaceRestRepository methods with authorizations checking

Commit:
DS-3542: Spring Authorization - @Preauthorization

Commit:
DS-3542: Fixes after rebase

tomdesair commented:
Hi @abollini,

I've restored the original methods on the `DSpaceRestRepository` class (see commit https://github.com/DSpace/DSpace/pull/1976/commits/798264c00174d445ce93d39bf381b6031f3ce2ba). I could not make the methods final as that prevents Spring Security to create a proxy for those methods. I've found another way to activate the Spring Security checks on internal class class by auto-wiring a self-reference in the repository.

I think that I've resolved all feedback from @abollini and @tdonohue and that this PR is ready to merge now.

abollini commented on dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/AuthorityEntryLinkRepository.java:
```diff
@@ -54,6 +55,7 @@ public HALResource wrapResource(AuthorityEntryRest model, String... rels) {
         return new AuthorityEntryResource(model);
     }
 
+    @PreAuthorize("hasAuthority('AUTHENTICATED')")
```
I'm not sure that this should be really restricted to authenticated user. It could be needed in the public UI to display richer information for metadata linked to an authority

tdonohue reviewed (approved):
👍 Gave this another code review. This looks good to me, all changes I suggested have been applied.

abollini reviewed (approved):
It looks good. IT should guarantee about correctness. 
I just added a inline comment about the security check on the authority endpoint, it is more for discussion so it can be merged as is.
It will be useful to check if the current angular app still works with these changes before to deploy the changes to the demo rest server

tdonohue commented:
Merging as this is at +3.
